scenario_player/utils/legacy.py

```diff
@@ -143,21 +153,149 @@ def reclaim_eth(
+        proxy_manager.client.wait_until_block(BlockNumber(ready_at_block + 20), retry_timeout=10)
+
+        log.info("Withdraw", user_address=address, amount=amount.__format__(",d"))
+        reclaim_amount += amount
+        userdeposit_proxy.withdraw(amount)
+
+    log.info(
+        "All UDC withdraws finished, now claim the resulting tokens!",
+        withdraw_total=reclaim_amount.__format__(",d"),
+    )
+
+    reclaim_erc20(
+        reclamation_candidates,
+        userdeposit_proxy.token_address("latest"),
+        contract_manager,
+        account,
+        eth_rpc_endpoint,
+    )
+
+
+def reclaim_erc20(
+    reclamation_candidates: Dict[ChecksumAddress, PrivateKey],
+    token_address: TokenAddress,
+    contract_manager: ContractManager,
+    account: Account,
+    eth_rpc_endpoint: URI,
+):
+    web3 = Web3(HTTPProvider(eth_rpc_endpoint))
 
-    log.info("Reclaiming candidates", addresses=list(address_to_keyfile.keys()))
+    reclaim_amount = 0
+
+    log.info("Checking chain")
+    for address, privkey in reclamation_candidates.items():
+        client = JSONRPCClient(web3, privkey)
+        token = CustomToken(client, token_address, contract_manager, "latest")
+
+        balance = token.balance_of(to_canonical_address(address))
+        log.debug(
+            "balance", token=to_checksum_address(token_address), balance=balance, address=address
+        )
+        if balance > 0:
+            drain_amount = balance
+            log.info(
+                "Reclaiming tokens", from_address=address, amount=drain_amount.__format__(",d")
+            )
+            reclaim_amount += drain_amount
+            assert account.address
+            try:
+                token.transfer(account.address, TokenAmount(drain_amount))
+            except InsufficientEth:
+                log.warning(
+                    "Not sufficient eth in node wallet to reclaim",
+                    address=address,
+                    token_address=token_address,
+                )
+                continue
+
+    log.info(
+        "Reclaimed", reclaim_amount=reclaim_amount.__format__(",d"), token_address=token.address
+    )
+
+
+def reclaim_eth(
+    reclamation_candidates: Dict[ChecksumAddress, PrivateKey],
+    account: Account,
+    eth_rpc_endpoint: URI,
+):
+    web3 = Web3(HTTPProvider(eth_rpc_endpoint))
 
     txs = []
     reclaim_amount = 0
     gas_price = web3.eth.gasPrice
     reclaim_tx_cost = gas_price * VALUE_TX_GAS_COST
 
     log.info("Checking chain")
-    for address, keyfile_content in address_to_keyfile.items():
+    for address, privkey in reclamation_candidates.items():
         balance = web3.eth.getBalance(address)
         if balance > reclaim_tx_cost:
-            if address not in address_to_privkey:
-                address_to_privkey[address] = decode_keyfile_json(keyfile_content, b"")
-            privkey = address_to_privkey[address]
```

Comment:
I'm thinking about placing a `reclaim_done` file into each node dir after the reclaim. This would avoid this performance problem and also allow us to skip the balance check for old nodes on each run. If you really want to rerun the reclaim for some reason, it's easy enough to delete that file. What do you think?

Reply:
Sounds like a good idea. If we then update the scenario run part to delete that file if it re-funds a node things should work out fine.

The question is if it still makes sense to delay the keyfile loading until it's necessary. But we can see how it performs and change it later if needed.